### ISSUE TITLE
Plugin Repository Structure Standardization

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#2419]
### Goal

To discuss and agree on a standard structure that Rock plug-in repositories could (should?) adhere to in order to simplify the sharing of code between developers.

### Background

Currently most plug-in developers have their own repository structure. This makes it difficult to quickly jump into another developers repo and understand where things are and how to find new things. It also presents difficulty in designing scraper services that could aggregate all plug-in development progress in one place so developers can quickly see what others are working on (see related #2418).

@nairdo @edmistj @azturner @mikejed @chrismcgrath @cbump @trydyingtolive @arranf @garrettjohnson @Tschrock

**Comment by cabal95** [comment]
Below is a writeup of how I currently structure my plugin repositories. Hopefully it makes sense. This is just one implementation currently in use by developers to be discussed.

### Folder Structure

Here is a brief outline of the folder structure that I use in my github
repositories. A sample of this structure format can be seen in this repo, https://github.com/ShepherdDev/digital-signage

```
|- com.organization.ProjectName.csproj
|- Controls\ (excluded from project file)
|- Documentation\ (1,2)
|- Migrations\ (1)
|- Rest\ (1)
|- SystemGuid\ (1)
|- Themes\ (excluded from project file)
|- Webhooks\ (excluded from project file)
|- Releases\
   \- ##PluginVersionNumber##\
      |- install.sql
      |- uninstall.sql
      |- installdelete.txt
      \- uninstalldelete.txt
```

1. These folders are not really part of any particular layout structure, they
are just examples of how I tend to organize things currently. For a cleaner
structure it might make sense to move all of these underneath a single `Code/`
folder or something similar.
2. The Documentation folder tends to contain just pictures that I link to from
the README.md or Wiki of the github project for screenshots and such.

### RockIt and Rock Production

There is a Powershell script in each repository that can be executed to setup
all the various hard links in the file system. When you execute the script it
will ask you to select either the RockIt folder or your production Rock folder.

> Note: Selecting the production folder is useful when developing/debugging
> against a git install of Rock, such as when you need to debug interactions
> with core DLLs.

You can run the script as many times as you need and it will only perform the
tasks that are needed. So if you add a new theme you can just re-run the script
and it will only link that one new theme you added.

The following hard links (junctions) are made:

* com.organization.ProjectName.csproj - Linked into your RockIt folder. This
helps keep project DLL references sane so that they can use a common path to
the core DLLs, such as `../RockWeb/bin/Rock.dll`.
* Controls/ - Becomes your plugin folder inside RockWeb, such as
`RockWeb/Plugins/com_organization/ProjectName/`. The `com_organization` and
`ProjectName` are automatically determined from the name of the csproj file.
* Themes/ - Each sub-folder is linked into the RockWeb/Themes/ folder. So if
you have a `Themes/MyCustomTheme` it will be linked as `RockWeb/Themes/MyCustomTheme`.
* Webhooks/ - Each .ashx file will be linked into the `RockWeb/Webhooks` folder in
a similar fashion as the Themes were.

This hard linking provides the benefit that when you open your RockIt solution,
each of your projects is visible as a root level entry. These projects would
only show/contain any DLL code that is needed by your plugin. If the plugin
provides nothing more than a simple Block/UserControl with no custom compiled
code then the project is essentially just a placeholder for you to edit the
Release information.

All your user controls, themes and webhooks are visible and edited from inside
the RockWeb project in the solution (see screenshot below). From your point of view, at this point you
are editing a monolithic repo where everything is inside a single RockIt
repository. The only real difference is that when you perform git operations
you will be performing them on the "real" repo folder for that project.

This setup has allowed me to keep a clean distinction between projects when
working in the git UI on my desktop. If I have modified files that are not yet
committed to the repo in one project, I won't see those outstanding changes when
working in a different project. But when working inside Visual Studio I still
see a nice "combined" view of everything together.

### Packaging

After setting up this initial repo layout (without the Releases folder) I
realized this layout would work well for automating the packaging of a package.

Each plugin release number in the folder structure would be associated with a
version number in the Rock plugin shop. This generally means a sequential
numbering such as 1, 2, 3, 4, etc. All four files in the version folder are
optional, but if provided they provide the following data (and are renamed
accordingly to match the plugin format):

* install.sql - The SQL file to execute when installing the plugin
* uninstall.sql - The SQL file to execute when uninstalling the plugin
* installdelete.txt - The files to be deleted when this plugin version is installed.
* uninstalldelete.txt - The files to be deleted when this plugin version is uninstalled.

In my repositories I have started adding in a second Powershell script that
handles the packaging of a release when it is ready. The "quick" overview of
what this script does is as follows:

1. Perform a release build of the project for the compiled code.
   1. If the build fails, then the package building process aborts.
   2. _Currently this happens even if it is not needed, such as no code to
   compile, so it can probably be improved upon by only performing this step if
   a standard folder that all compiled code goes in exists, such as Code, or BLL,
   or whatever._
2. A staging folder of `_Package` is created to copy everything into so that we
can later zip it all up.
3. Any Blocks controls are copied into the staging directory.
4. Any Themes are copied into the staging directory.
5. Any Webhooks are copied into the staging directory.
6. The DLL built in step 1 is copied into the staging directory.
   1. _I think this also copies any dependency DLLs, but I can't remember for
   sure. If it doesn't we might need a way to "hint" to the build script which
   additional DLLs need to be copied in._
7. SQL install/uninstall scripts are copied into the staging directory.
8. Any extra files are copied into the staging directory.
   1. Currently this is the two `deletefile.lst` files if they exist.
9. Generate a list of files in the staging directory's content folder and Append
that to any user-supplied `deletefile.lst`.
10. Compress the staging folder into a .plugin zip file and delete the staging
files.
11. Move the built .plugin package into the proper Release version folder.

Many of those steps above only execute if they need to, specifically steps 3,
4, 5, 7 and 8. If you have not defined any SQL scripts for that release version
number then no SQL files are copied, and so on.

Step 9 will automatically generate the list of files to be deleted on uninstall,
but you need to provide the install(upgrade) file list for deletion. i.e. if you
rename a .CSS file in version 2, you need to provide the old file name in the
`installdelete.txt` file. The new file name will automatically be recorded in
the uninstall list for you. If for some reason there are files that need to be
deleted that the packager doesn't know about, such as generating CSS files from
LESS files at run-time, you can still manually indicate those in the
`uninstalldelete.txt` file.

Step 11 is because I like to keep built products in my repo. It's not really
best practice but the .plugin files are small enough and now once I have
"version 3" packaged, tested and ready for release I can do a final commit and
tag the repo so that I can always get back to the same source _and_ compiled
code.

### Version Numbers

So, to release a new version with the build script you simply create a new `##`
folder inside the `Releases` folder. So if you are getting ready to release
version 3 you would create a `Releases/3` folder and begin putting in any SQL
scripts or whatever else you need for that version number. At this point each
time you run the `BuildPackage.ps1` script it will build version 3 for you. If
you realize you forgot something or when you test it notice that something
didn't work the way you expected, you can make changes and re-run the script
and it will rebuild version 3 again.

### Sample Visual Studio Layout

Below is a screenshot from Visual Studio showing my local RockIt installation. I have highlighted a few things that are actually hard links to other locations on disk, even though they appear to be inside the RockIt solution. A few of the locations are actually slightly different (for example our website theme is not actually part of digital-signage, but you get the idea).

<img width="720" alt="sample rockit" src="https://user-images.githubusercontent.com/1119863/29895529-3fcce07c-8d8e-11e7-84e1-84f9f247cfc7.png">

### Potential Issues

There is only one potential known issue I know of, but have not tested to see if
it really would be a problem. Branches.

If you make use of branches some of the hard links _may_ get out of sync when
switching between branches. It depends on how git handles the process of switching
branches. The only files that would be affected by this for minor branch switches
are Webhooks and the .csproj file itself. If git fully deletes the files and
re-creates them then the hard link would likely still reference the "old" file
(that's just the way hard links work). If git simply opens the file and changes
all the contents to what they need to be then it should still work.

If this is indeed a problem we could make some tweaks to the Powershell script
so that it more gracefully handles this by removing any old hard links and
creating new hard links for everything each time it is run.

If you use branching heavily, especially if you use a branch for non-code
related files (such as a `documentation` branch of `website` branch for
github.io or whatever it is called) then it could cause heavy problems with the
hard links, as the entire `Controls` folder might vanish. In this case I would
suggest either checking out two copies of the repository (one for code, one for
the "other" branches) or an alternative repository structure all together.

I tend to not use branching heavily on these project repositories thus far. The
times I have it has usually been a simple branching. I'm making edits on a
control, not done yet but want to commit the code. Since I don't know if I'm
even going to keep the changes (not fully tested) I don't want to commit to the
master branch, so I make a new branch to commit those changes to. These types
of branching have not caused any problems for me yet.

There is the possibility of using symbolic links instead of hard links to resolve
these issues. I don't remember if I had tried that and ran into trouble
or if I just did hard links "just because". I'll see if I can test and report
on the functionality of symbolic links in IIS, IISExpress and Visual Studio.

**Comment by edmistj** [comment]
This is super helpful. I'll try to play with this soon. I was thinking it would be nice if we had a standard set of badges for Github for things like 'In Rock Shop' or 'Stable' (perhaps others).

**Comment by cabal95** [comment]
I just did some testing and switched one of my projects to use symbolic links instead of hard/junction links and everything seems to work fine.  IIS Express traverses them just fine. Visual Studio sees them just fine. IIS on Server 2008R2 sees them just fine. So symbolic links should work fine and take care of the potential issues listed above.

However (of course). Windows does not let you create symbolic links unless you are the administrator. Doesn't matter if you are in in the Administrators group. Somehow, my Windows 10 Dev VM does let me do so, but my regular windows 10 desktop does not. Windows Server 2008, 2012 and 2016 also do not. This might be why I switched to Hard links and Junctions because it will let you do that (rolling eyes).

The script could be modified to do symbolic links and use UAC to elevate privileges to Administrator when doing so.
